src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                    wasDeletedOnServer = true;
+                }
+                auto otherJob = _discoveryData->_queuedDeletedDirectories.take(originalPath);
+                if (otherJob) {
+                    delete otherJob;
+                    wasDeletedOnServer = true;
+                }
+
+                auto postProcessRename = [this, item, base, originalPath](PathTuple &path) {
+                    auto adjustedOriginalPath = _discoveryData->adjustRenamedPath(originalPath);
+                    _discoveryData->_renamedItems.insert(originalPath, path._target);
+                    item->_modtime = base._modtime;
+                    item->_inode = base._inode;
+                    item->_instruction = CSYNC_INSTRUCTION_RENAME;
+                    item->_direction = SyncFileItem::Down;
+                    item->_renameTarget = path._target;
+                    item->_file = adjustedOriginalPath;
+                    item->_originalFile = originalPath;
+                    path._original = originalPath;
+                    path._local = adjustedOriginalPath;
+                    qCInfo(lcDisco) << "Rename detected (down) " << item->_file << " -> " << item->_renameTarget;
+                };
+
+                if (wasDeletedOnServer) {
+                    postProcessRename(path);
+                    done = true;
+                } else {
+                    // we need to make a request to the server to know that the original file is deleted on the server
+                    _pendingAsyncJobs++;
+                    auto job = new RequestEtagJob(_discoveryData->_account, originalPath, this);
+                    connect(job, &RequestEtagJob::finishedWithResult, this, [=](const Result<QString> &etag) mutable {
+                        _pendingAsyncJobs--;
+                        QTimer::singleShot(0, _discoveryData, &DiscoveryPhase::scheduleMoreJobs);
+                        if (etag.errorCode() != 404 ||
+                            // Somehow another item claimed this original path, consider as if it existed
+                            _discoveryData->_renamedItems.contains(originalPath)) {
+                            // If the file exist or if there is another error, consider it is a new file.
+                            postProcessServerNew();
+                            return;
+                        }
+
+                        // The file do not exist, it is a rename
+
+                        // In case the deleted item was discovered in parallel
+                        auto it = _discoveryData->_deletedItem.find(originalPath);
+                        if (it != _discoveryData->_deletedItem.end()) {
+                            ASSERT((*it)->_instruction == CSYNC_INSTRUCTION_REMOVE);
+                            (*it)->_instruction = CSYNC_INSTRUCTION_NONE;
+                        }
+                        delete _discoveryData->_queuedDeletedDirectories.take(originalPath);
```

Comment:
This check-local, async-check, check-local-again pattern is a bit awkward but looks correct.

One worry: since the async job is started before done is true, could there be multiple deletion-check jobs running at the same time? Then when they finish both could consume the rename?

done=true before going async? (depends on how friendly we want to be towards duplicate fileids)